RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc

```diff
@@ -0,0 +1,141 @@
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/MuonReco/interface/Muon.h"
+#include "DataFormats/MuonReco/interface/MuonFwd.h"
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidate.h"
+#include "DataFormats/ParticleFlowCandidate/interface/PFCandidateFwd.h"
+
+//
+// class declaration
+//
+
+class BadChargedCandidateFilter : public edm::global::EDFilter<> {
+public:
+  explicit BadChargedCandidateFilter(const edm::ParameterSet&);
+  ~BadChargedCandidateFilter();
+
+private:
+  virtual bool filter(edm::StreamID iID, edm::Event&, const edm::EventSetup&) const override;
+
+      // ----------member data ---------------------------
+
+  edm::EDGetTokenT<edm::View<reco::Candidate> >   tokenPFCandidates_;
+  edm::EDGetTokenT<edm::View<reco::Muon> >   tokenMuons_;
+
+  const bool taggingMode_;
+  const bool debug_;
+  const double          maxDR_;
+  const double          minPtDiffRel_;
+  const double          minMuonTrackRelErr_;
+  const double          minMuonPt_;
+
+};
+
+//
+// constructors and destructor
+//
+BadChargedCandidateFilter::BadChargedCandidateFilter(const edm::ParameterSet& iConfig)
+  : tokenPFCandidates_ ( consumes<edm::View<reco::Candidate> >(iConfig.getParameter<edm::InputTag> ("PFCandidates")  ))
+  , tokenMuons_ ( consumes<edm::View<reco::Muon> >(iConfig.getParameter<edm::InputTag> ("muons")  ))
+  , taggingMode_          ( iConfig.getParameter<bool>    ("taggingMode") )
+  , debug_                ( iConfig.getParameter<bool>    ("debug") )
+  , maxDR_                ( iConfig.getParameter<double>  ("maxDR") )
+  , minPtDiffRel_         ( iConfig.getParameter<double>  ("minPtDiffRel") )
+  , minMuonTrackRelErr_   ( iConfig.getParameter<double>  ("minMuonTrackRelErr") )
+  , minMuonPt_            ( iConfig.getParameter<double>  ("minMuonPt") )
+{
+  produces<bool>();
+}
+
+BadChargedCandidateFilter::~BadChargedCandidateFilter() { }
+
+
+//
+// member functions
+//
+
+// ------------ method called on each new Event  ------------
+bool
+BadChargedCandidateFilter::filter(edm::StreamID iID, edm::Event& iEvent, const edm::EventSetup& iSetup) const
+{
+  using namespace std;
+  using namespace edm;
+
+  typedef View<reco::Candidate> CandidateView;
+  Handle<CandidateView> pfCandidates;
+  iEvent.getByToken(tokenPFCandidates_,pfCandidates);
+
+  typedef View<reco::Muon> MuonView;
+  Handle<MuonView> muons;
+  iEvent.getByToken(tokenMuons_,muons);
+
+  bool foundBadChargedCandidate = false;
+
+  for ( unsigned i=0; i < muons->size(); ++i ) { // loop over all muons
+
+    const reco::Muon & muon = (*muons)[i];
+    if (debug_) cout<<"BadChargedCandidate test"<<endl;
```

Comment:
Note that "cout" is not thread safe and will produce garbled output if the code is running multithreaded.
Better change to "LogInfo" (not printed by MessageLogger by default) or "LogPrint" (printed by MessageLogger by default) to use thread safe version.
We typically allow the pattern ``` if (debug) cout << ... ```, but it should be understood that it has issues and will not allow debugging of a job running with multiple threads.